File: packages/asdl-core/src/asdl_core/payloads/root.py
"""Payload root and session-id resolution helpers."""

from __future__ import annotations

import os
import tempfile
from collections.abc import Mapping
from pathlib import Path

from asdl_core.payloads.errors import PayloadError
from asdl_core.payloads.segments import is_safe_segment

ASDL_PAYLOAD_ROOT_ENV = "ASDL_PAYLOAD_ROOT"
ASDL_PAYLOAD_SESSION_ID_ENV = "ASDL_PAYLOAD_SESSION_ID"


def default_payload_root(*, temp_dir: Path | None = None) -> Path:
    """Return the default payload root under the platform temp directory."""

    base_temp_dir = temp_dir if temp_dir is not None else Path(tempfile.gettempdir())
    return base_temp_dir / "asdl"


def resolve_payload_root(
    *,
    env: Mapping[str, str] | None = None,
    temp_dir: Path | None = None,
) -> Path:
    """Resolve the payload root from ``ASDL_PAYLOAD_ROOT`` or the default temp root."""

    source_env = os.environ if env is None else env
    env_value = source_env.get(ASDL_PAYLOAD_ROOT_ENV)
    if env_value is None or env_value == "":
        return default_payload_root(temp_dir=temp_dir)

    payload_root = Path(env_value)
    if payload_root.is_absolute():
        return payload_root
    raise PayloadError(
        error_type="payload_root_invalid",
        message=f"{ASDL_PAYLOAD_ROOT_ENV} must be an absolute path: {env_value!r}",
    )


def resolve_payload_session_id(
    explicit_session_id: str | None = None,
    *,
    env: Mapping[str, str] | None = None,
) -> str:
    """Resolve and validate a supplied payload session id."""

    source_env = os.environ if env is None else env
    if explicit_session_id is not None and explicit_session_id != "":
        session_id = explicit_session_id
    else:
        env_session_id = source_env.get(ASDL_PAYLOAD_SESSION_ID_ENV)
        if env_session_id is None or env_session_id == "":
            raise PayloadError(
                error_type="payload_session_required",
                message=(
                    "Payload artifact mode requires a session id from an explicit option "
                    f"or {ASDL_PAYLOAD_SESSION_ID_ENV}."
                ),
            )
        session_id = env_session_id

    if is_safe_segment(session_id):
        return session_id
    raise PayloadError(
        error_type="payload_session_invalid",
        message=f"Payload session id must be a safe segment: {session_id!r}",
    )

File: ts/packages/aretro/src/payloads/root.ts
/**
 * Payload root and session-id resolution helpers.
 */

import * as os from "node:os";
import * as path from "node:path";

import { PayloadError } from "./errors.ts";
import { isSafeSegment } from "./segments.ts";

export const ASDL_PAYLOAD_ROOT_ENV = "ASDL_PAYLOAD_ROOT";
export const ASDL_PAYLOAD_SESSION_ID_ENV = "ASDL_PAYLOAD_SESSION_ID";

export function defaultPayloadRoot(options?: { tempDir?: string }): string {
	const baseTempDir = options?.tempDir ?? os.tmpdir();
	return path.join(baseTempDir, "asdl");
}

export function resolvePayloadRoot(options?: {
	env?: Record<string, string | undefined>;
	tempDir?: string;
}): string {
	const sourceEnv = options?.env ?? process.env;
	const envValue = sourceEnv[ASDL_PAYLOAD_ROOT_ENV];
	if (envValue === undefined || envValue === "") {
		return defaultPayloadRoot(
			options?.tempDir !== undefined ? { tempDir: options.tempDir } : undefined,
		);
	}

	if (path.isAbsolute(envValue)) {
		return envValue;
	}
	throw new PayloadError(
		"payload_root_invalid",
		`${ASDL_PAYLOAD_ROOT_ENV} must be an absolute path: ${JSON.stringify(envValue)}`,
	);
}

export function resolvePayloadSessionId(
	explicitSessionId: string | undefined,
	options?: { env?: Record<string, string | undefined> },
): string {
	const sourceEnv = options?.env ?? process.env;
	let sessionId: string;
	if (explicitSessionId !== undefined && explicitSessionId !== "") {
		sessionId = explicitSessionId;
	} else {
		const envSessionId = sourceEnv[ASDL_PAYLOAD_SESSION_ID_ENV];
		if (envSessionId === undefined || envSessionId === "") {
			throw new PayloadError(
				"payload_session_required",
				`Payload artifact mode requires a session id from an explicit option or ${ASDL_PAYLOAD_SESSION_ID_ENV}.`,
			);
		}
		sessionId = envSessionId;
	}

	if (isSafeSegment(sessionId)) {
		return sessionId;
	}
	throw new PayloadError(
		"payload_session_invalid",
		`Payload session id must be a safe segment: ${JSON.stringify(sessionId)}`,
	);
}
